designs/2025-define-plugin/README.md

```diff
@@ -0,0 +1,628 @@
+- Repo: eslint/rewrite
+- Start Date: 2025-04-23
+- RFC PR: <https://github.com/eslint/rfcs/pull/132>
+- Authors: [JoshuaKGoldberg](https://github.com/JoshuaKGoldberg)
+
+# Exporting a `definePlugin` function
+
+## Summary
+
+Proposes a `definePlugin` function to `@eslint/plugin-kit` as the recommended way to group plugin configs, meta, and rules.
+
+## Motivation
+
+Defining the exported values of an ESLint plugin is a "loose" process right now.
+ESLint core provides only recommendations, not standardized factory functions; the community also has centralized only on common patterns.
+Several dozen variations in how plugins are defined have evolved, ranging from manual declarations of `configs`, to custom scripts, to auto-generated JavaScript or TypeScript files.
+
+This divergence in implementation details results in conceptual and practical overhead for plugin authors.
+Without a standardized factory function to define a plugin:
+
+- There is no first-party way for plugins to receive type hints during development the way [`eslint/config`'s `defineConfig`](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file) assists in authoring configs
+- Plugin names are often defined in at least four separate places:
+  - `configs[string].name`, such as `"example/recommended"`
+  - `configs[string].rules`, such as `{ "example/rule": "error" }`
+  - `rules[string]`, such as `{ "example/rule": { ... } }`
```

Comment:
When defining the rules objects for a plugin, it doesn't seem like the names are prefixed here. https://eslint.org/docs/latest/extend/plugins#configs-in-plugins

This would just be `{ "rule": { ... } }` and then the `example/` would be set when the config registered this plugin as `"example": pluginObject`.

It would be convenient for this utility to have some way of either automatically prefixing these rules when defining the config, or otherwise enforce that all of the rules exist.

Reply:
Agreed, that would be nice. It would be nice if the new `rules` could take in an array of rules objects.

The tricky point is, what would assign the rule name for the rules then? Rule `meta` objects don't have a name field in ESLint core. typescript-eslint's `RuleCreator` does but that's non-standard.

I would personally be in favor of:

* Adding an optional `meta.name: string` property
* `defineConfig` allowing an array of rules rather than object _iff_ each has a unique `meta.name`

Reply:
I don't think there's anything wrong with having to define the `rules` as a map of "non-namespaced name" to rule object. It's when we have to add the namespace to specify the rules in the configs that this gets potentially messy.

I like the approach of having `meta.docs.recommended` to determine the `recommended` config, but perhaps that could go all the way to it's conclusion: each rule could locally define (`meta`) which configs it should be in?

Reply:
> each rule could locally define (`meta`) which configs it should be in?

Yeah! That's what I'm going for with `meta.docs.recommended`. The first version of this RFC just has `true` support for a `recommended`. From feedback in https://github.com/eslint/rfcs/pull/132/files#r2059307675 I'm expanding it now to also support valid configs. I suspect going further than that would get into tricky territory. Some plugins define specific configs as supersets of others, and having to describe all of those in each individual rule's `meta` is a lot of duplicate & heavy work.